lib/classes/cart_checkout.rb

```diff
@@ -0,0 +1,94 @@
+#
+# Helper class for the checkout process
+#
+# Process:
+#   1. designate each package to the order
+#   2. clear the cart
+#   3. submit the order
+#
+# Usage:
+#   CartCheckout
+#     .designate_cart_items(items, <opts>)
+#     .to_order(order)
+#
+# Options:
+#   - ignore_unavailable {boolean}
+#     > if true, will checkout regardless of whether some packages are unavailable or not
+#
+# Notes:
+#   Currently only for singleton items
+#
+class CartCheckout
+  include ::ActiveModel::Validations
+
+  attr_accessor :cart_items
+
+  # --- Entry point
+
+  def self.designate_cart_items(cart_items, ignore_unavailable: false)
+    CartCheckout.new(cart_items, ignore_unavailable: ignore_unavailable)
+  end
+
+  # --- Action
+
+  def to_order(order)
+    steps = [
+      -> { validate(order) },
+      -> { add_cart_items_to_order(order) },
+      -> { submit_order(order) }
+    ]
+
+    steps.each do |step|
+      step.call()
+      break if errors.any?
+    end
+    errors
+  end
+
+  private
+
+  def initialize(cart_items, ignore_unavailable: false)
```

Comment:
I will recommend initialize method not to keep as private. Purpose of the class is we can use it from anywhere else as well. If you still want to keep it's scope limited to inheritance then will recommend making it protected. But I will suggest making it public.

Mostly initialize method written as the first method in class some typical ways which ruby people follow. 😸

Reply:
This is similar to a singleton pattern, we do *not* want the users to initialize it manually.
The api of this class is made to be used with the `designate_cart_items` entry point.

Reply:
It's by design

Reply:
👍